<a href="https://colab.research.google.com/github/RubenMcCarty/Desicion_Making_Model_UPN/blob/main/Lab_06_Data_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data Wrangling o Negociacion de datos**
## [MSc:Ruben Quispe](https://www.linkedin.com/in/msc-rub%C3%A9n-quispe-l/)

* Manejar valores perdidos
* formatos correctos de datos
* Estandarizar y normalizar datos

# Contenido
* Identificar y manejar Valores perdidos (missing values)
    * Indentifico valores perdidos
    * Resolver los valores perdidos 
    * Formatos de datos correcto
* Estandarización de Datos
* Normalización de datos
* Binning
* Variable Indicadora




# ¿Qué es el data wrangling? o cuál es el propósito de MANIPULACIÓN DE DATOS?

Data Wrangling es el proceso de convertir datos del formato inicial a un formato que puede ser mejor para el análisis.

¿Cuál es la tasa de consumo de combustible (L / 100k) para el automóvil diesel?

# Importar data

In [1]:
import pandas as pd
import matplotlib.pylab as plt
import numpy as np

In [2]:
data = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"

In [4]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [5]:
df = pd.read_csv(data, names = headers)

In [6]:
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

Como podemos ver, aparecieron varios signos de interrogación en el marco de datos; esos son valores perdidos que pueden dificultar nuestro análisis posterior.

Entonces, ¿cómo identificamos todos esos valores perdidos y los tratamos?
¿Cómo trabajar con datos faltantes?

Pasos para trabajar con datos faltantes:

identificar los datos faltantes
lidiar con los datos faltantes
formato de datos correcto

In [9]:
df.replace('?', np.nan, inplace=True)
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

# Evaluación de datos faltantes 
Los valores faltantes se convierten a los predeterminados. Usamos las siguientes funciones para identificar estos valores faltantes. Hay dos métodos para detectar datos faltantes:
*  .isnull()
*  .notnull()

In [10]:
missing_data = df.isnull()
missing_data.head()

symboling  normalized-losses   make  ...  city-mpg  highway-mpg  price
0      False               True  False  ...     False        False  False
1      False               True  False  ...     False        False  False
2      False               True  False  ...     False        False  False
3      False              False  False  ...     False        False  False
4      False              False  False  ...     False        False  False

[5 rows x 26 columns]

In [11]:
missing_data

symboling  normalized-losses   make  ...  city-mpg  highway-mpg  price
0        False               True  False  ...     False        False  False
1        False               True  False  ...     False        False  False
2        False               True  False  ...     False        False  False
3        False              False  False  ...     False        False  False
4        False              False  False  ...     False        False  False
..         ...                ...    ...  ...       ...          ...    ...
200      False              False  False  ...     False        False  False
201      False              False  False  ...     False        False  False
202      False              False  False  ...     False        False  False
203      False              False  False  ...     False        False  False
204      False              False  False  ...     False        False  False

[205 rows x 26 columns]

"TRUE" significa valor perdido, mientras que "FALSE" significa valor no perdido.

# Cuente los valores perdidos en cada columna
Usando un bucle for en Python, podemos calcular rápidamente el número de valores faltantes en cada columna. Como se mencionó anteriormente, "TRUE" representa un valor perdido, "FALSE" significa que el valor está presente en el conjunto de datos. En el cuerpo del bucle for, el método ".value_counts ()" cuenta el número de valores "TRUE".




In [18]:
for column in missing_data.columns.values.tolist():
  print(column)
  print(missing_data[column].value_counts())
  print('__________________________________________')


symboling
False    205
Name: symboling, dtype: int64
__________________________________________
normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64
__________________________________________
make
False    205
Name: make, dtype: int64
__________________________________________
fuel-type
False    205
Name: fuel-type, dtype: int64
__________________________________________
aspiration
False    205
Name: aspiration, dtype: int64
__________________________________________
num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64
__________________________________________
body-style
False    205
Name: body-style, dtype: int64
__________________________________________
drive-wheels
False    205
Name: drive-wheels, dtype: int64
__________________________________________
engine-location
False    205
Name: engine-location, dtype: int64
__________________________________________
wheel-base
False    205
Name: wheel-base, dtype: int64
_____________________

# Según el resumen anterior, cada columna tiene 205 filas de datos, siete columnas que contienen datos faltantes:

1. "normalized-losses": 41 missing data
2. "num-of-doors": 2 missing data
3. " bore": 4 missing data
4. "stroke" : 4 missing data
5. "horsepower": 2 missing data
6. "peak-rpm": 2 missing data
7. "price": 4 missing data

# Hacer frente o resolver los datos faltantes
* ¿Cómo lidiar con los datos faltantes?
#1. Eliminar (dropear) datos
  * a. Dropear toda la fila
  * b. Dropear toda la columna
#2. reemplazar datos
  * a. reemplazarlo por la media
  * b. reemplazarlo por frecuencia
  * c. reemplazarlo en función de otras funciones

Las columnas enteras deben descartarse solo si la mayoría de las entradas de la columna están vacías. En nuestro conjunto de datos, ninguna de las columnas está lo suficientemente vacía como para eliminarse por completo. Tenemos cierta libertad para elegir qué método reemplazar los datos; sin embargo, algunos métodos pueden parecer más razonables que otros. Aplicaremos cada método a muchas columnas diferentes:

# Reemplazar por media:

* "normalized-losses": 41 datos faltantes, reemplácelos con la media
* "stroke": 4 datos faltantes, reemplácelos con la media
* "bore": 4 datos faltantes, reemplácelos con mean
* "horsepower": 2 datos faltantes, reemplácelos con la media
* "peak-rpm": 2 datos faltantes, reemplácelos con la media

# Reemplazar por frecuencia:

"num-of-doors": faltan 2 datos, reemplácelos con "cuatro".
Razón: el 84% de los sedanes son de cuatro puertas. Dado que cuatro puertas es más frecuente, es más probable que ocurra

# Drop toda la fila:

* "price": 4 datos faltantes, simplemente elimine toda la fila
Razón: el precio es lo que queremos predecir. Cualquier entrada de datos sin datos de precios no se puede utilizar para la predicción; por lo tanto, cualquier fila ahora sin datos de precios no nos es útil

# Calcular el promedio o media de la columna NORMALIZED LOSSES

In [19]:
missing_data

symboling  normalized-losses   make  ...  city-mpg  highway-mpg  price
0        False               True  False  ...     False        False  False
1        False               True  False  ...     False        False  False
2        False               True  False  ...     False        False  False
3        False              False  False  ...     False        False  False
4        False              False  False  ...     False        False  False
..         ...                ...    ...  ...       ...          ...    ...
200      False              False  False  ...     False        False  False
201      False              False  False  ...     False        False  False
202      False              False  False  ...     False        False  False
203      False              False  False  ...     False        False  False
204      False              False  False  ...     False        False  False

[205 rows x 26 columns]

In [16]:
avg_norm_loss = df['normalized-losses']. astype(float).mean(axis =0)
print('La media de normalized-losses:', avg_norm_loss)

La media de normalized-losses: 122.0


# Reemplazar NaN por el valor media en la columna 'normalized-losses'

In [22]:
df['normalized-losses'].replace(np.nan, avg_norm_loss, inplace = True)
df

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3               122  alfa-romero  ...       21          27  13495
1            3               122  alfa-romero  ...       21          27  16500
2            1               122  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

# Calcular la media la columna bore

In [24]:
avg_bore = df['bore']. astype(float).mean(axis =0)
print('La media de bore:', avg_bore)

La media de bore: 3.3297512437810957


# Reemplazar NaN por el valor de la media 

In [26]:
df['bore'].replace(np.nan, avg_bore, inplace=True)
df

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3               122  alfa-romero  ...       21          27  13495
1            3               122  alfa-romero  ...       21          27  16500
2            1               122  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

# De acuerdo a lo explicado antes, reemplazar NaN in la columna 'stroke' por la media

In [27]:
avg_stroke = df['stroke'].astype('float').mean(axis = 0)
print('el promedio de stroke es:', avg_stroke)

el promedio de stroke es: 3.2554228855721337


In [28]:
df['stroke'].replace(np.nan, avg_stroke, inplace = True)

In [29]:
df

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3               122  alfa-romero  ...       21          27  13495
1            3               122  alfa-romero  ...       21          27  16500
2            1               122  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]